### PR TITLE
Inception Score Implementation

### PR DESCRIPTION
**Pull request by ahmedo42**
Related Issue #998 

Description:

To reproduce inception score on CIFAR10 check this notebook

https://colab.research.google.com/drive/1odZBAzltKl6OmNuA7zsnkeXLsh5YPEhi?usp=sharing

References:

- https://github.com/sbarratt/inception-score-pytorch
- https://github.com/NVlabs/stylegan2-ada-pytorch/blob/main/metrics/inception_score.py

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

**Comment by sdesrozis** [comment]
@ahmedo42 Thank you for this PR! As usual, I suppose that @vfdev-5 will be very fast to review but I will do my best too since this topic is very interesting to me.

File: ignite/contrib/handlers/neptune_logger.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ class NeptuneLogger(BaseLogger):
     Args:
         api_token: Required in online mode. Neputne API token, found on https://neptune.ai.
             Read how to get your API key
-            https://docs.neptune.ai/administration/security-and-privacy/how-to-find-and-set-neptune-api-token.
+            https://docs.neptune.ai/integrations-and-supported-tools/ide-and-notebooks/jupyter-lab-and-jupyter-notebook/notebooks-1#where-is-neptune_api_token.
```
Comment (vfdev-5) [inline ignite/contrib/handlers/neptune_logger.py:45]: @ahmedo42 let's revert this change as I removed those links in the master.

File: ignite/metrics/gan/inception_score.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        self.n_splits = splits
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._probs = []  # type: List[torch.Tensor]
+
+    @reinit__is_reduced
+    def update(self, output: torch.Tensor) -> None:
+        generated = output.detach()
+        inception_output = self.inception_model(generated)
```
Comment (vfdev-5) [inline ignite/metrics/gan/inception_score.py:95]: We should add `with torch.no_grad()` context manager here to avoid grads creation.

File: ignite/metrics/gan/inception_score.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            _probs_tensor = cast(torch.Tensor, idist.all_gather(_probs_tensor))
+        self._is_reduced = True
+
+        result = 0.0
+        if idist.get_rank() == 0:
+            N = _probs_tensor.shape[0]
+            scores = torch.zeros((self.n_splits,))
+            for i in range(self.n_splits):
+                part = _probs_tensor[i * (N // self.n_splits) : (i + 1) * (N // self.n_splits)]
+                kl = part * (torch.log(part) - torch.log(torch.mean(part, dim=0)))
```
Comment (vfdev-5) [inline ignite/metrics/gan/inception_score.py:116]: Please, if we could reuse pytorch functions to compute KL. We can also probably avoid applying softmax in the `update` pass.

File: ignite/metrics/gan/inception_score.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                kl = torch.mean(torch.sum(kl, dim=1))
+                scores[i] = torch.exp(kl)
+            result = torch.mean(scores).item()
```
Comment (vfdev-5) [inline ignite/metrics/gan/inception_score.py:119]: Can't we accumulate `kl` on the `update` to avoid saving all `probs` ?

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @ahmedo42 
I think we can have this metric with online implementation.

File: ignite/metrics/gan/inception_score.py (hunk 4 above)
Comment (ahmedo42) [reply ignite/metrics/gan/inception_score.py:119]: If what you meant is this:
```python
    @reinit__is_reduced
    def update(self, output: torch.Tensor) -> None:
        generated = output.detach()
        with torch.no_grad():
            inception_output = self.inception_model(generated)
        probs = self._apply_softmax(inception_output)
        kl = self._compute_kl(probs)
        self._kl += kl
        self._num_examples += probs.shape[0]

    @sync_all_reduce("_kl","_num_examples")
    def compute(self) -> float:
        return self._kl.item() / self._num_examples
```
I think this is not valid. 

The problem is how to incorporate `splits` into the calculations , when you have the whole dataset it's easy to split it into well-defined segments ,calculate the `kl` for each and then average .

Comment (vfdev-5) [reply ignite/metrics/gan/inception_score.py:119]: @ahmedo42 you can think of `n_splits` of KL accumulators. Here is an examples with counter (which you can replace with KL) : 
```python
n_splits = 12
batch_size = 32
num_samples = 2000

n_iters = (num_samples + batch_size - 1) // batch_size 

# reset():
iteration = 0
counter = torch.zeros(n_splits)

for j in range(n_iters):
    
    # update():
    for i in range(batch_size):
        counter[iteration % n_splits] += 1
        iteration += 1

> 
> counter
> tensor([168., 168., 168., 168., 168., 168., 168., 168., 168., 168., 168., 168.])
```


Have you also seen this paper: https://arxiv.org/abs/1801.01973 where they suggest to get rid of `n_splits` ?

Comment (ahmedo42) [reply ignite/metrics/gan/inception_score.py:119]: @vfdev-5  thanks for the idea , this is the implementation based on what you said , the inception score from this is 10.3 (compared to 9.6 of the previous implementation under the same conditions):
```python

    @reinit__is_reduced
    def update(self, output: torch.Tensor) -> None:
        generated = output.detach()
        with torch.no_grad():
            inception_output = self.inception_model(generated)
        probs = self._apply_softmax(inception_output)
        batch_size = probs.shape[0]
        for i in range(batch_size):
            self._kl[self._iteration % self.n_splits] += self._compute_kl(probs[i])
            self._iteration += 1
        self._num_examples += probs.shape[0]

    @sync_all_reduce("_kl","_num_examples")
    def compute(self) -> float:
        return torch.mean(self._kl).item() / self._num_examples

    def _compute_kl(self,input_tensor:torch.Tensor):
        kl = input_tensor * (torch.log(input_tensor) - torch.log(torch.mean(input_tensor, dim=0)))
        kl = torch.sum(kl, dim=0)
        return torch.exp(kl)
```
I saw the paper and I think we should provide the original implementation ( as it's still widely used now ) , then we can extend that.

Comment (vfdev-5) [reply ignite/metrics/gan/inception_score.py:119]: I think this implementation is incorrect due to `_compute_kl` : 
- `torch.log(torch.mean(input_tensor, dim=0))` originally computes the average on `num_samples // n_splits` of tensors and in your case it is batch size
-  `torch.exp(kl)` should not be applied here.

**Comment by vfdev-5** [comment]
@ahmedo42 do you plan to continue working on this PR ?

**Comment by ahmedo42** [comment]
@vfdev-5 Sorry about the absence , I've been busy with college and will probably be for the next two weeks ( midterms ) , Iet's close it and then reopen it I think .

**Comment by vfdev-5** [comment]
Closing the PR in favor of #2053
